### PR TITLE
fix: support SIP-018 for structured data signing

### LINKED ISSUE #2419: Signed messages should return a signature that has recoverable public key
Talking with @pradel about how to use the new signed messages feature in thing like ceramic.network, we realized that the signatures the wallet is currently sending down to the apps are ones that make it impossible to recover the public key from the signature. 

The wallet is using `signECDSA` versus `signWithKey`.

Sign with Key allows for recoverable public keys from signatures:
```ts
export function signWithKey(privateKey: StacksPrivateKey, input: string): MessageSignature {
  const [rawSignature, recoveryParam] = signSync(input, privateKey.data.slice(0, 32), {
    canonical: true,
    recovered: true,
  });
  const signature = Signature.fromHex(rawSignature);
  const coordinateValueBytes = 32;
  const r = leftPadHexToLength(signature.r.toString(16), coordinateValueBytes * 2);
  const s = leftPadHexToLength(signature.s.toString(16), coordinateValueBytes * 2);

  if (recoveryParam === undefined || recoveryParam === null) {
    throw new Error('"signature.recoveryParam" is not set');
  }
  const recoveryParamHex = intToHexString(recoveryParam, 1);
  const recoverableSignatureString = recoveryParamHex + r + s;
  return createMessageSignature(recoverableSignatureString);
}
```

Where as `signECDSA` does not:

```ts
export function signECDSA(
  privateKey: string,
  content: string | Buffer
): {
  publicKey: string;
  signature: string;
} {
  const contentBuffer = content instanceof Buffer ? content : Buffer.from(content);
  const publicKey = getPublicKeyFromPrivate(privateKey);
  const contentHash = hashSha256Sync(contentBuffer);
  const signature = signSync(contentHash, privateKey);

  return {
    signature: utils.bytesToHex(signature),
    publicKey,
  };
}
```

We can also see that Marvins example implementation correctly uses signWithKey https://github.com/MarvinJanssen/stx-signed-structured-data/blob/4c5714e2b9957db0037863d15c3640a0524abe20/scripts/deps.ts#L28

I'd suggested shipping a change to where you switch out these functions.

### LINKED ISSUE #2443: Hash displayed in wallet for signed message is incorrect
Was pulling my hair out trying to figure out why my hashes were different, yet was able to validate the signatures lol

https://github.com/hirosystems/stacks-wallet-web/blob/23d9d9ac388ecf54f570be93f0b909cc662f0b2e/src/app/pages/signature-request/components/message-box.tsx#L24-L27

this bit here is incorrect, should be using the shared hashMessage function.

cc @kyranjamie @beguene

### LINKED ISSUE #2460: signStructuredData not in line with SIP018 specification
See https://github.com/hirosystems/connect/issues/225#issue-1245926927

As part of this task, we need to make sure that all the signatures used during the signing flow are in RSV format.

References
https://github.com/stacksgov/sips/pull/57
https://github.com/hirosystems/stacks.js/pull/1269
https://github.com/hirosystems/connect/issues/225

### LINKED ISSUE #2463: Ensure account and network respected when signing message
When attempting to sign a message, the network or stxAddress passed to the wallet don't seem to make any difference -- I tested the test app you are using in this repo and it seems to be the case there too.

I'd say this is a P1-2 bug, should be base functionality imo. The network does not matter much, but it's confusing to the end user. The bigger issue is it's likely users won't notice that the incorrect address is being used to sign (if they are using an address that is not index 0), and would then sign a message with a different account than they were intending for.

cc @kyranjamie @beguene

### PR DESCRIPTION
**Pull request by beguene**
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/stacks-wallet-web/actions/runs/2548029860).

This PR implements the updated version of SIP-018 https://github.com/stacksgov/sips/pull/57
and fixes the following too

* return all signatures in RSV format
* display correct hash in the UI

closes #2443
closes #2460
closes #2419
closes #2463

### CODE REVIEW

**Comment by beguene** [comment]
@landitus Now the structured message also has a `domain` value like this
```javascript
  const domain = tupleCV({
    name: stringAsciiCV('hiro.so'),
    version: stringAsciiCV('1.0.0'),
    'chain-id': uintCV(1),
  }); 
```
domain

I added a message and domain header, but we could do a bit better. Let me know if you have something better.

File: package.json
Comment (kyranjamie) [inline package.json]: This isn't necessary, right?

File: package.json
Comment (fbwoolf) [inline package.json]: This version has been released ...same with `connect-react` below.

File: src/app/pages/signature-request/signature-request.tsx
Comment (kyranjamie) [inline src/app/pages/signature-request/signature-request.tsx]: Is there a cleaner way to do this, that doesn't involve conditional setting of empty vars? Should we be using two different components for string/structured message signing.

File: src/app/pages/signature-request/components/message-box.tsx
Comment (fbwoolf) [inline src/app/pages/signature-request/components/message-box.tsx]: I don't think a Stack is needed here bc there is only one element?

File: test-app/src/components/signature.tsx
Comment (fbwoolf) [inline test-app/src/components/signature.tsx:125]: Are you intentionally leaving these logs? There is one on L125 too.

EDIT: I see now this is in the test-app so you likely mean to leave them. I'll resolve.

**Review by fbwoolf — APPROVED** [review]

**Comment by MarvinJanssen** [comment]
This (very) old concept drawing displayed the app name and version on the top. Maybe something like that is useful for the domain object? Additionally, perhaps it could show a suffix of `(mainnet)` or `(testnet)` based on the `chain-id` property.

SIP018 display concept

File: package.json
Comment (beguene) [reply package.json]: This is not the same as the released v6.8.4. This packages is actually https://github.com/hirosystems/connect/pull/236. I will update it when we release my connect PR.
Same for connect-react.

File: package.json
Comment (beguene) [reply package.json]: Yes indeed. I will revert it.

File: src/app/pages/signature-request/signature-request.tsx
Comment (beguene) [reply src/app/pages/signature-request/signature-request.tsx]: Sure I will think of something else.

**Comment by beguene** [comment]
thanks @MarvinJanssen I will try something like that.

**Comment by beguene** [comment]
updated popup
structured1

File: src/app/pages/signature-request/signature-request.tsx
Comment (beguene) [reply src/app/pages/signature-request/signature-request.tsx]: I split into components to make it clearer.

File: package.json
Comment (kyranjamie) [reply package.json]: this hasn't been reverted yet

**Comment by markmhendrickson** [comment]
@beguene is this awaiting a design integration review from @landitus?

**Comment by beguene** [comment]
@markmhx No I am working on those test unit issues
https://github.com/hirosystems/stacks-wallet-web/runs/6985575450?check_suite_focus=true#step:5:825

The upgrade to the latest connect-ui v5.5.1 fails many tests. Those `connect` changes are not related to AMS but AMS support in connect is built on top of it, so the upgrade is necessary.
This should be the last thing pending for this PR.

**Comment by beguene** [comment]
@janniks I upgraded our connect to the latest version which includes your latest changes https://github.com/hirosystems/connect/pull/234
It's failing many unit tests in the web-wallet via jest.
https://github.com/hirosystems/stacks-wallet-web/runs/6985575450?check_suite_focus=true#step:5:825
It looks like node_modules/@stacks/connect/dist/connect.cjs.development.js:13:20 is loading connect-modal.js via import instead of loading the connect-modal.cjs file which is not co-located next to the .js but is in the cjs folder instead.
I am trying to find a solution for the web wallet with some custom jest config
like
```
   moduleNameMapper: {
     '@stacks/connect-ui/dist/components/connect-modal':
       '<rootDir>/node_modules/@stacks/connect/node_modules/@stacks/connect-ui/dist/cjs/connect-modal.cjs.entry.js',
  },
```
without success.

It seems that `connect` should package it with the right import rather than having to change our web-wallet jest config.
I saw that you are working on this https://github.com/hirosystems/connect/pull/238
Is it related ? Is there a solution that I can use right now ?

**Comment by janniks** [comment]
Yep, I think I'll revert the change and add as a separate module. Will ping you later with progress.

**Comment by beguene** [comment]
@timstackblock This is ready for QA. There are special button in test-app to test the different types of message.
cc: @markmhx @andresgalante

**Comment by beguene** [comment]
@fbwoolf There is nothing left for me to do on this PR. I will let you merge it whenever you need it.

**Comment by fbwoolf** [comment]
@beguene do all of the linked issues need to be QA'd by me here before merging or do you feel confident it can be merged and then tested with a release PR?

**Comment by beguene** [comment]
@fbwoolf It's up to you, you can maybe quickly try it, there are some buttons I added in the test-app.

**Comment by landitus** [comment]
> updated popup structured1

Sorry I missed this @beguene! Yep, those headers will do the trick just fine 👍

**Pull request merged**
